Public/Src/Utilities/Storage/FileContentInfo.cs

```diff
@@ -18,27 +18,153 @@ namespace BuildXL.Storage
+            // The top 8 bits are used for storing PathExistence value, the rest 56 bits are used to store a file length.
+            // The top in each of these block is used to indicate whether the value is specified.            
+            // 
+            // m_data = [IsKnownExistenceFlag | Existence | IsKnownLengthFlag | Length]; 
+            //
+            // Note: because the left most bit in m_data might be set, don't right shift its value without casting it to ulong or unsetting the bit first.
+
+            private const int NumberExistenceBits = 8;
+
+            private const int ExistenceShift = 64 - NumberExistenceBits;
+
+            private const long LengthBitsMask = (1L << ExistenceShift) - 1;
+            private const long IsKnownLengthFlag = 1L << (ExistenceShift - 1);
+            private const long LengthValueMask = (1L << (ExistenceShift - 1)) - 1;
+
+            private const long ExistenceBitsMask = ~LengthBitsMask;
+            private const long IsKnownExistenceFlag = 1L << 63;
+            private const long ExistenceValueMask = ExistenceBitsMask ^ IsKnownExistenceFlag;            
+
+            private readonly long m_data;
+
+            /// <summary>
+            /// Max file length that can be stored.
+            /// </summary>
+            public const long MaxSupportedLength = LengthValueMask;
+
+            /// <summary>
+            /// Whether the length value has been set.
+            /// </summary>
+            public bool IsKnownLength => (m_data & IsKnownLengthFlag) != 0;
+
+            /// <summary>
+            /// Length value. Returns 0 if <see cref="IsKnownLength" /> is false.
+            /// </summary>
+            public long Length => m_data & LengthValueMask;
+
+            /// <summary>
+            /// The underlying value used for storage.
+            /// </summary>
+            public long CombinedValue => m_data;
+
+            /// <summary>
+            /// The stored existence value. If no value is stored, returns null.
+            /// </summary>
+            public PathExistence? Existence
+            {
+                get
+                {
+                    if ((m_data & IsKnownExistenceFlag) != 0)
+                    {
+                        return (PathExistence)((m_data & ExistenceValueMask) >> ExistenceShift);
+                    }
+
+                    return null;
+                }
+            }
+
+            /// <summary>
+            /// Creates a <see cref="LengthAndExistence"/> with a known length and optional existence value.
+            /// </summary>
+            public LengthAndExistence(long length, PathExistence? existence)
+            {
+                if ((length & LengthValueMask) != length)
+                {
+                    Contract.Assert(false, $"Invalid length value (reserved bits should not be used): {length} ({System.Text.RegularExpressions.Regex.Replace(Convert.ToString(length, 2).PadLeft(64, '0'), ".{4}", "$0 ")})");

```

Comment:
>"Invalid length value (reserved bits should not be used): {length} ({System.Text.RegularExpressions.Regex.Replace(Convert.ToString(length, 2).PadLeft(64, '0'), ".{4}", "$0 ")})"); [](start = 44, length = 179)

Would it be simpler to just have a message that "length cannot be larger than x"?

Reply:
Well, the main purpose is to check that someone is not passing a previously serialized value. The condition is essentially checking that length is within [0..MaxSupportedLength]

---
In reply to: [301852095](https://github.com/microsoft/BuildXL/pull/570#discussion_r301852095) [](ancestors = 301852095)